benchmark/paddle/CARLA_SAC/train.py

```diff
@@ -0,0 +1,194 @@
+    action_dim = env.action_space.shape[0]
+
+    # Initialize model, algorithm, agent, replay_memory
+    model = CarlaModel(obs_dim, action_dim)
+    algorithm = SAC(
+        model,
+        gamma=GAMMA,
+        tau=TAU,
+        alpha=ALPHA,
+        actor_lr=ACTOR_LR,
+        critic_lr=CRITIC_LR)
+    agent = CarlaAgent(algorithm)
+    rpm = ReplayMemory(
+        max_size=MEMORY_SIZE, obs_dim=obs_dim, act_dim=action_dim)
+
+    total_steps = 0
+    test_flag = 0
+    obs_list = [env.reset() for env in env_list]
+    obs_list = [obs.get() for obs in obs_list]
+    obs_list = np.array(obs_list)
+    done_list = [False] * len(env_list)
+    episode_reward_list = [0] * len(env_list)
+    episode_steps_list = [0] * len(env_list)
+
+    while total_steps < args.max_timesteps:
+        # Train episode
+        if rpm.size() < WARMUP_STEPS:
+            action_list = [
+                np.random.uniform(-1, 1, size=action_dim)
+                for _ in range(len(env_list))
+            ]
+        else:
+            action_list = [agent.sample(obs) for obs in obs_list]
+        return_list = [
+            env_list[i].step(action_list[i]) for i in range(len(env_list))
+        ]
+        return_list = [return_.get() for return_ in return_list]
+        return_list = np.array(return_list, dtype=object)
+
+        done_list = return_list[:, 2]
+        # Store data in replay memory
+        for i in range(len(return_list)):
+            next_obs, reward, done, _ = return_list[i, 0], return_list[
+                i, 1], return_list[i, 2], return_list[i, 3]
+            rpm.append(obs_list[i], action_list[i], reward, next_obs, done)
+
+            obs_list[i] = return_list[i, 0]
+
+            total_steps += 1
+            episode_steps_list[i] += 1
+            episode_reward_list[i] += reward
+            if done or episode_steps_list[i] >= _max_episode_steps:
+                tensorboard.add_scalar('train/episode_reward_env{}'.format(i),
+                                       episode_reward_list[i], total_steps)
+                logger.info('Total Steps: {} Reward: {}'.format(
+                    total_steps, episode_reward_list[i]))
+
+                # reset env if done
+                episode_steps_list[i] = 0
+                episode_reward_list[i] = 0
+                obs_list_i = env_list[i].reset()
+                obs_list[i] = obs_list_i.get()
+                obs_list[i] = np.array(obs_list[i])
+
+        # Train agent after collecting sufficient data
+        if rpm.size() >= WARMUP_STEPS:
+            batch_obs, batch_action, batch_reward, batch_next_obs, batch_terminal = rpm.sample_batch(
+                BATCH_SIZE)
+            agent.learn(batch_obs, batch_action, batch_reward, batch_next_obs,
+                        batch_terminal)
+
+        # Save agent
+        if total_steps > int(1e5) and total_steps % int(1e4) == 0:
+            agent.save('./model/{}_model.ckpt'.format(total_steps))
```

Comment:
./model/step_{}_model.ckpt

Reply:
if total_steps > int(1e5) and total_steps > last_save_steps + int(1e4):
            agent.save('./model/step_{}_model.ckpt'.format(total_steps))
            last_save_steps = total_steps